### PR TITLE
dmd: set template recursion limit via CLI, instead of being hardcoded to 500

### PR DESCRIPTION
**Pull request by dteod**

### CODE REVIEW

**Comment by UplinkCoder** [comment]
`v` stands for verbose.
and turns on informational messages.
you should not have the switch start with v.

**Comment by dteod** [comment]
> `v` stands for verbose.
> and turns on informational messages.
> you should not have the switch start with v.

Ok, thank you.
Removed it and alphabetically reordered.

File: src/dmd/globals.h [hunk 1]
```diff
@@ -172,6 +172,7 @@ struct Param
 
     CHECKACTION checkAction;       // action to take when bounds, asserts or switch defaults are violated
 
+    unsigned recursionLimit;
```
Comment (ibuclaw) [inline src/dmd/globals.h:175]: I think you have to remove the old enum too? I might be thinking of the cxx branch though.

File: src/dmd/cli.d [hunk 2]
```diff
@@ -645,6 +645,9 @@ dmd -cov -unittest myprog.d
                 to the file $(TT profilegc.log) upon program termination.)
             )`,
         ),
+	Option("recursion-limit=<num>",
```
Comment (ibuclaw) [inline src/dmd/cli.d:648]: It would be better to have `template` in the option name.

File: src/dmd/mars.d [hunk 3]
```diff
@@ -1777,6 +1777,23 @@ bool parseCommandLine(const ref Strings arguments, const size_t argc, ref Param
...
+	    if (p[STRING_SIZE] != '=')
+	    {
+		errorInvalidSwitch(p, "Expected argument following `-recursion-limit , e.g. `-recursion-limit=1000`");
+		return true;
+	    }
+	    if (!params.recursionLimit.parseDigits(p.toDString()[STRING_SIZE + 1 .. $]))
+	    {
+                errorInvalidSwitch(p, "Only numbers are allowed for -recursion-limit`");
+	    }
+	    if( params.recursionLimit == 0)
```
Comment (ibuclaw) [inline src/dmd/mars.d:1792]: Styling needs adjusting here.

File: src/dmd/cli.d (hunk 2 above)
Comment (ibuclaw) [reply src/dmd/cli.d:648]: Ahh, but it also controls recursive type depth, so template may not be such an accurate description afterall.

File: src/dmd/mars.d (hunk 3 above)
Comment (dteod) [reply src/dmd/mars.d]: You are right, I am sorry (vim tabs misleaded me).
Corrected.

File: src/dmd/globals.h (hunk 1 above)
Comment (dteod) [reply src/dmd/globals.h:175]: In the last commit I removed the enum from frontend.h and replaced it, based on the Param::errorLimit variable that has a similar behavior

File: src/dmd/cli.d (hunk 2 above)
Comment (dteod) [reply src/dmd/cli.d:648]: Removed the inaccuracy

**Comment by dteod** [comment]
I was implementing an (hobbyist) assembler for ARM in D, the idea was to add an enumeration with 506 instruction types, templatize a struct to be instantiated with a mixin according to the syntax of the instruction and merge them together in an Algebraic (or SumType) to return from a `parse` function (something like `alias AnyInstruction = mixin(getAnyInstructionDeclarationString()); AnyInstruction parse(string);` ).
Unfortunately the recursion depth limit of 500 made impossible for me to implement it this way, so I changed the implementation to something more sustainable (and normal lol) like a single struct holding all the possible variations of the various ARM instructions. It works, but each instantiation of Instruction will take more memory.

After this, I came up with the modification. The feature is a simple setting switch over a default value and is not invasive for anyone's workflow. Also it could be useful for linear algebra programmers using CTFE, compile-time generators writers and various others. gcc and clang for C++ also have a similar switch (-fconstexpr-depth, -ftemplate-depth) and I managed to use it sometimes.

**Comment by ibuclaw** [comment]
This will need a changelog entry to be written up about the added switch.

**Comment by dteod** [comment]
> This will need a changelog entry to be written up about the added switch.

Added the changelog with a small example of the use case this PR is trying to solve. Please let me know if there are any other issues or even suggestions

Thank you

**Comment by ibuclaw** [comment]
> > This will need a changelog entry to be written up about the added switch.
> 
> Added the changelog with a small example of the use case this PR is trying to solve. Please let me know if there are any other issues or even suggestions
> 
> Thank you

Based off the Codecov warnings, there should be tests added to check the new switch, both accepted and failure paths.

**Comment by RazvanN7** [comment]
Since @WalterBright is usually against adding new switches, we need his approval on this one.

**Comment by dteod** [comment]
> Since @WalterBright is usually against adding new switches, we need his approval on this one.

I understand. I did the modification locally and I thought it could be useful for the user to extend D's usage on corner cases upon request. I chose to add a switch in order to be the least invasive as possible.

**Comment by dteod** [comment]
> > > This will need a changelog entry to be written up about the added switch.
> > 
> > 
> > Added the changelog with a small example of the use case this PR is trying to solve. Please let me know if there are any other issues or even suggestions
> > Thank you
> 
> Based off the Codecov warnings, there should be tests added to check the new switch, both accepted and failure paths.

Hello, it took some time but eventually I added the tests in the latest commits.

**Comment by dteod** [comment]
> > > This will need a changelog entry to be written up about the added switch.
> > 
> > 
> > Added the changelog with a small example of the use case this PR is trying to solve. Please let me know if there are any other issues or even suggestions
> > Thank you
> 
> Based off the Codecov warnings, there should be tests added to check the new switch, both accepted and failure paths.

Hello, it took some time but eventually I added the tests in the latest commits.

**Comment by RazvanN7** [comment]
The tests are still failing because you add a newline in the headers:

```
     Array<Identifier* >* debugids;
     bool hasMainFunction;
     enum : int32_t { recursionLimit = 500 };
+
     uint32_t startGagging();
     bool endGagging(uint32_t oldGagged);
     void increaseErrorCount();
```

@WalterBright do you approve this addition?

File: src/dmd/frontend.h [hunk 4]
```diff
@@ -856,7 +859,6 @@ struct Global final
     Array<Identifier* >* debugids;
     bool hasMainFunction;
     enum : int32_t { recursionLimit = 500 };
-
```
Comment (dteod) [inline src/dmd/frontend.h:859]: @RazvanN7 Looks like this line is not present anymore

**Comment by dteod** [comment]
> The tests are still failing because you add a newline in the headers:
> 
> ```
>      Array<Identifier* >* debugids;
>      bool hasMainFunction;
>      enum : int32_t { recursionLimit = 500 };
> +
>      uint32_t startGagging();
>      bool endGagging(uint32_t oldGagged);
>      void increaseErrorCount();
> ```
> 
> @WalterBright do you approve this addition?

Looks like the offending line is not present anymore, but [the failing test](https://dev.azure.com/dlanguage/dmd/_build/results?buildId=24376&view=logs&j=2e8e4583-3493-5411-a5ed-6353b3fd9e1d&t=1ff6c36c-9a21-5e8b-abe8-4f4b2e4baa8c&l=3978) states
`{'detail': ErrorDetail(string='Unable to locate build via Azure API. Project is likely private, please upload with the Codecov repository upload token to resolve issue.', code='not_found')}`

**Comment by thewilsonator** [comment]
> Codecov repository

Ignore those errors they are unrelated. Rebasing might help.

File: src/dmd/frontend.h (hunk 4 above)
Comment (dteod) [reply src/dmd/frontend.h:859]: Misunderstood the `diff`, the like should be present

**Comment by dteod** [comment]
> > Codecov repository
> 
> Ignore those errors they are unrelated. Rebasing might help.

Thank you, I sync'd the repos and crossed the fingers

**Comment by thewilsonator** [comment]
Ah, ignore that, I thought https://github.com/dlang/dmd/pull/13092 was merged already, its not.  Ignore those coverage errors for the moment.

**Comment by dteod** [comment]
> Ah, ignore that, I thought #13092 was merged already, its not. Ignore those coverage errors for the moment.

I discovered it just 5 minutes ago, I'll follow and rebase when it will be merged

**Comment by thewilsonator** [comment]
> when it will be merged

Its merged

**Comment by dteod** [comment]
> > when it will be merged
> 
> Its merged

All right, sync'd the forked repo

**Comment by dteod** [comment]
Hello, could someone approve the last test please?
I removed a leftover of the old recursionLimit enum in order to prevent potential bugs in future code

**Comment by dteod** [comment]
> Sigh. Did anyone stop to ask why this was needed? Even @dteod wrote: "so I changed the implementation to something more sustainable (and normal lol)". Does this sound like a feature we need?
> 
> The limit was put in because no limit meant weird errors would result from stack overflows.
> 
> I'm disappointed this was pulled without any pushback at all on _why_.
> 
> > Since @WalterBright is usually against adding new switches, we need his approval on this one.
> 
> This PR illustrates why. It should be reverted. The compiler has too many switches already, adding these kinds of options just makes it harder and harder to understand.

Personally I wanted to extend D's usage on particular corner cases that couldn't be managed by the default implementation... One of the reasons I adopted D was because of CTFE, that is similar in fashion to template metaprogramming (that in C++ can be controlled via a switch) and it has in common the benefit of avoiding to spread repositories with files containing compile-time data (Excel files, mainly).
The switch is just an optional addition to the default behavior to let the user have more control on the D interpreter.

**Comment by WalterBright** [comment]
> I really don't understand why the end user of the language shall be limited in this way,

Because end users get very unhappy with stack overflow faults. Another problem is they may not realize their code is recursing like this.

> and why D is not welcoming to accept compiler hacks even for small things like these.

Every switch increases the difficulty for users to understand how to use the compiler. They increase the testing required. They move the compiler further away from "just working". They make it harder to understand someone else's makefile. Makefiles accumulate compiler switches like barnacles on a ship, and the people who inherit it generally have no idea why those switches were used.

> The user doesn't have be a genius at all to understand what the switch does... whatever

If it was the only switch, yes, but more and more switches lead to a ball of mud that nobody understands. Hence each switch needs to have a really good case for it, and this one does not. You wrote that you wrote a better version of your algorithm that didn't need this - great!

**Pull request merged**